src/cascade/executor/estimate_location.py

```diff
@@ -52,24 +51,68 @@ def estimate_location(execution_context, local_settings, local_cache=None):
     covariate_multipliers, covariate_data_spec = create_covariate_specifications(
         local_settings.settings.country_covariate, local_settings.settings.study_covariate
     )
+    local_cache.set("covariate_multipliers:{local_settings.parent_location_id}", covariate_multipliers)
+    local_cache.set("covariate_data_spec:{local_settings.parent_location_id}", covariate_data_spec)
     input_data = retrieve_data(execution_context, local_settings, covariate_data_spec, local_cache)
     columns_wrong = validate_input_data_types(input_data)
     assert not columns_wrong, f"validation failed {columns_wrong}"
     modified_data = modify_input_data(input_data, local_settings, covariate_data_spec)
+    local_cache.set("prepared_input_data:{local_settings.parent_location_id}", modified_data)
+
+
+def construct_model_for_estimate_location(local_settings, local_cache):
+    covariate_multipliers = local_cache.get("covariate_multipliers:{local_settings.parent_location_id}")
```

Comment:
I don't like this get/set on the cache at all. Looks like we don't know how to write a function that has inputs and outputs. Maybe these functions should look like non side-effecting functions, and it's outside of these that we serialize and deserialize what they need.

Reply:
Oh, duh. This is the outside part. I see. And you've written the serialization by hand, which is an excellent first pass.

Reply:
Yeah, I agree this is ugly. You need _some_ way to annotate which values are passed through the cache. Could be function argument names and a returned dictionary where the outer wrapper takes care of packing and unpacking, that's more fiddly but makes this code look cleaner. Or you can do something like this where you explicitly pack and unpack values. I don't know what the best answer is. The explicit packing and unpacking doesn't bother me particularly. What does bother me is that, as I wrote this out, it felt like I was sticking stuff into the cache pretty arbitrarily, which I was. I think you could sit and think a little more carefully about how this get's partitioned and sequenced and make it feel more orderly.

Reply:
These messages are part of a protocol. In order to know bandwidth, to ensure there isn't crazy cache use, I like to have a list of what those things are or will be. Dynamic is good, but this is a nice place to put a layer. I'll think about it. Meanwhile, what you have is a good step.